compiler/rustc_mir_build/src/builder/expr/into.rs

```diff
@@ -365,6 +365,121 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
+                        // `write_box_via_move(b, val)` becomes
+                        // ```
+                        // *box_uninit_as_mut_ptr(&mut b) = val;
+                        // b
+                        // ```
+                        let tcx = this.tcx;
+                        let t = generic_args.type_at(0);
+                        let [b, val] = **args else {
+                            span_bug!(expr_span, "invalid init_box_via_move call")
+                        };
+                        let box_ty = this.thir.exprs[b].ty;
+                        let Some(b) = unpack!(block = this.as_local_operand(block, b)).place()
+                        else {
+                            span_bug!(expr_span, "invalid init_box_via_move call")
+                        };
+
+                        // Create reference to `b`.
+                        let box_ref = this
+                            .local_decls
+                            .push(LocalDecl::new(Ty::new_mut_ptr(tcx, box_ty), expr_span));
+                        this.cfg.push(
+                            block,
+                            Statement::new(source_info, StatementKind::StorageLive(box_ref)),
+                        );
+                        // Make sure `StorageDead` gets emitted.
+                        this.schedule_drop_storage_and_value(
+                            expr_span,
+                            this.local_scope(),
+                            box_ref,
+                        );
+                        this.cfg.push_assign(
+                            block,
+                            source_info,
+                            box_ref.into(),
+                            Rvalue::RawPtr(RawPtrKind::Mut, b),
+                        );
+
+                        // Invoke `box_uninit_as_mut_ptr`.
+                        let as_mut_ptr_fn = Operand::function_handle(
+                            tcx,
+                            tcx.require_lang_item(LangItem::BoxUninitAsMutPtr, expr_span),
+                            [t.into()],
+                            expr_span,
+                        );
+                        let inner_ptr = this.temp(Ty::new_mut_ptr(tcx, t), expr_span);
+                        let success = this.cfg.start_new_block();
+                        this.cfg.terminate(
+                            block,
+                            source_info,
+                            TerminatorKind::Call {
+                                func: as_mut_ptr_fn,
+                                args: [Spanned {
+                                    node: Operand::Copy(box_ref.into()),
+                                    span: expr_span,
+                                }]
+                                .into(),
+                                destination: inner_ptr,
+                                target: Some(success),
+                                unwind: UnwindAction::Continue,
```

Comment:
The `UnwindAction::Continue` here seems sus to me, don't we have to potentially insert some cleanup blocks to drop stuff that's locally in scope?

I copied this from the old box_new lowering.

Reply:
`box_uninit_as_mut_ptr` cannot unwind, so `UnwindAction::Unreachable` should be even better.

Reply:
Strangely that changes very little, the MIR building test for this shows that we still generate an unwinding path.